### ISSUE TITLE
TrialRunners need separate workingDirectory under Windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2482]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2482|
|Type|defect|
|Created|2007-02-25 20:19:49Z|

In the resolution of [#1901](https://github.com/twisted/twisted/issues/1901), we found a strange behaviour with trial under Windows leading to this error:

```
[ERROR]: twisted.trial.test.test_runner.TestTrialRunner.test_logFileGetsClosed Traceback (most recent call last):
File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\trial\test\test_runner.py", line 154, in test_logFileGetsClosed self.runner.run(self.test)
File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\trial\runner.py", line 791, in run oldDir = self._setUpTestdir()
File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\trial\runner.py", line 701, in _setUpTestdir os.path.abspath("_trial_temp_old%s"
exceptions.OSError: [Errno 13] Permission denied
```

exarkun suggested it could come from _tearDownLogging calls.

Attachments:

* [test_runner_2482.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/eaf/eaf7c6894787be4f764f36d44669301798770040/7d32fb9491dbeb6c5e64fab9aa667d8ce91e3a57.diff) (1856 bytes) - added by therve on 2007-03-01 18:05:28Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2482 2482
type__defect defect
reporter__therve therve
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1172434789000000 1172434789000000
changetime__1175759576000000 1175759576000000
version__None None
owner__ 
cc__exarkun cc__jml cc__teratorn
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

It seems it could not come from trial itself, but more of the way we use it inside the test. It seems it's required to call _tearDownLogFile, or some strange behavior occurs.

The patch attaches correct the problem. To reproduce before, run twisted.trial.test.test_runner with the '-u' flag.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @jml|
|-|-|

It seems likely to me that this cleanup code actually belongs in the runner somewhere.

Let's see if jml agrees.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

As a gut reaction, we should be moving as much stuff out of the runner as possible. I'll investigate further asap.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml set owner to @therve|
|-|-|

exarkun is right. `_tearDownLogFile` should be called in the finally clause following `suite.run()` in `TrialRunner.run`.

Please move it there, adjust the tests accordingly and re-assign to me for review. 

Also, if you feel inclined, please add a docstring for `test_runUntilFailure` and a docstring or a comment for `_test`.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @jml|
|-|-|

It seems good, it passes for the window buildsalve.

Ready to review in trial-teardownlogfile-2482.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml set owner to @therve|
|-|-|

Good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `closed`|
|-|-|

(In [19961]) Merge trial-teardownlogfile-2482

Author: therve
Reviewer: jml
Fixes [#2482](https://github.com/twisted/twisted/issues/2482)

Add a _tearDownLogFile call in the trial Runner run method, to cleanup log
directory. It's required for having a sane behaviour under Windows.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
